### PR TITLE
feat: add WindowedReader

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR provides an implementation of windowed table scan interface so that we can scan a small section of SST files and memtables to get the most recent records and sort according to sort options.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/storage/src/chunk.rs [hunk 1]
```diff
@@ -206,13 +242,13 @@ impl ChunkReaderBuilder {
 
     /// Check if SST file's time range matches predicate.
     #[inline]
-    fn file_in_range(file: &FileHandle, predicate: TimestampRange) -> bool {
-        if predicate == TimestampRange::min_to_max() {
+    fn file_in_range(file: &FileHandle, predicate: &TimestampRange) -> bool {
```
Comment (evenyag) [inline src/storage/src/chunk.rs:245]: Maybe we can remove the inline hint now

File: src/storage/src/chunk.rs [hunk 2]
```diff
@@ -151,51 +154,84 @@ impl ChunkReaderBuilder {
...
+    pub fn time_windows(mut self, windows: Vec<TimestampRange>) -> Self {
+        self.time_windows = Some(windows);
+        self
+    }
 
-        let schema = Arc::new(
-            ProjectedSchema::new(self.schema, self.projection)
-                .context(error::InvalidProjectionSnafu)?,
-        );
+    pub async fn build_windowed(
```
Comment (evenyag) [inline src/storage/src/chunk.rs:163]: Does it need to be pub? Seems we only call build()

File: src/storage/src/chunk.rs [hunk 3]
```diff
@@ -151,51 +154,84 @@ impl ChunkReaderBuilder {
...
+        schema: &ProjectedSchemaRef,
+        time_range: &TimestampRange,
+    ) -> Result<BoxedBatchReader> {
         let num_sources = self.memtables.len() + self.files_to_read.len();
         let mut reader_builder = MergeReaderBuilder::with_capacity(schema.clone(), num_sources)
             .batch_size(self.iter_ctx.batch_size);
 
-        self.iter_ctx.projected_schema = Some(schema.clone());
-        for mem in self.memtables {
+        for mem in &self.memtables {
```
Comment (evenyag) [inline src/storage/src/chunk.rs:189]: The memtable might also need to support scan by time range.

File: src/storage/src/read/windowed.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        Self { schema, readers }
+    }
+}
+
+#[async_trait::async_trait]
+impl<R> BatchReader for WindowedReader<R>
+where
+    R: BatchReader,
+{
+    async fn next_batch(&mut self) -> error::Result<Option<Batch>> {
```
Comment (evenyag) [inline src/storage/src/read/windowed.rs:46]: Import crate::error::Result?
```suggestion
    async fn next_batch(&mut self) -> Result<Option<Batch>> {
```

File: src/storage/src/read/windowed.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        let vectors_in_batch = existing.into_iter().map(|mut b| b.to_vector()).collect();
+        let res = sort_by_rows(&self.schema, vectors_in_batch)?;
+        Ok(Some(Batch::new(res)))
+    }
+}
+
+fn sort_by_rows(
+    schema: &ProjectedSchemaRef,
+    vectors: Vec<VectorRef>,
+) -> crate::error::Result<Vec<VectorRef>> {
```
Comment (evenyag) [inline src/storage/src/read/windowed.rs:73]:
```suggestion
) -> Result<Vec<VectorRef>> {
```

File: src/storage/src/read/windowed.rs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            .iter()
+            .map(|c| c.desc.data_type.create_mutable_vector(32))
+            .collect::<Vec<Box<dyn MutableVector>>>();
+
+        while let Some(batch) = reader.next_batch().await? {
+            for (builder, arr) in existing.iter_mut().zip(batch.columns.iter()) {
+                builder
+                    .extend_slice_of(&**arr, 0, arr.len())
+                    .context(error::PushBatchSnafu)?;
+            }
```
Comment (evenyag) [inline src/storage/src/read/windowed.rs:61]: Could we collect all the vectors and finally use [concat](https://docs.rs/arrow/latest/arrow/compute/fn.concat.html) to merge them?

File: src/storage/src/read/windowed.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+fn sort_by_rows(
+    schema: &ProjectedSchemaRef,
+    vectors: Vec<VectorRef>,
+) -> crate::error::Result<Vec<VectorRef>> {
+    let sort_columns = build_sort_columns(schema);
+
+    let store_schema = schema.schema_to_read();
+    // Convert columns to rows to speed lexicographic sort
+    let mut row_converter = RowConverter::new(
```
Comment (evenyag) [inline src/storage/src/read/windowed.rs:78]: Seems arrow provides [lexsort_to_indices](https://docs.rs/arrow/latest/arrow/compute/fn.lexsort_to_indices.html), could we use it?

Comment (v0y4g3r) [reply src/storage/src/read/windowed.rs:78]: > Seems arrow provides [lexsort_to_indices](https://docs.rs/arrow/latest/arrow/compute/fn.lexsort_to_indices.html), could we use it?

`lexsort_to_indices` sorts data by columns, while it's prefered to sort data by rows to improve locality as mentioned in [this article](https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-2/).

Comment (evenyag) [reply src/storage/src/read/windowed.rs:78]: Get it. I also have this concern when implementing the merge reader.

**Review by waynexia — COMMENTED** [review]
Mostly LGTM

File: src/storage/src/read/windowed.rs (hunk 7 above)
Comment (waynexia) [inline src/storage/src/read/windowed.rs:78]: We can add a special branch for the scenario that only sort on the time index column.

Comment (v0y4g3r) [reply src/storage/src/read/windowed.rs:86]: I added a todo here to remind me to optimize sort with only timestamp column involved.

**Comment by killme2008** [comment]
@v0y4g3r Please add some description for this PR.

File: src/storage/src/read/windowed.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            batches.push(
+                batch
+                    .columns
+                    .into_iter()
+                    .map(|v| v.to_arrow_array())
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        let num_columns = batches.get(0).map(|b| b.len()).unwrap_or(0);
```
Comment (killme2008) [inline src/storage/src/read/windowed.rs:61]: Adds a fast path when the `batches` is empty?

File: src/storage/src/read/windowed.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+    debug_assert_eq!(sorted.len(), store_schema.num_columns());
+
+    Ok(sorted)
+}
+
+/// [<PK_1>, <PK_2>, TS] to [TS, <PK_1>, <PK_2>]
+fn build_sort_columns(schema: &ProjectedSchemaRef) -> Vec<(usize, bool)> {
+    let row_key_end = schema.schema_to_read().row_key_end();
+    let ts_col_index = row_key_end - 1;
```
Comment (killme2008) [inline src/storage/src/read/windowed.rs:131]: I think we can add a function `time_column_index` for `StoreSchema` to get the time column index. We should not hard code the implicit time index position of row keys in other places. It's not maintainable.

File: src/storage/src/chunk.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/storage/src/chunk.rs:189]: This PR depends on #1576 so that we can filter timestamps in memtable.

**Review by waynexia — APPROVED** [review]
🚀

File: src/storage/src/memtable/btree.rs [hunk 10]
```diff
@@ -196,25 +204,53 @@ impl BTreeIterator {
...
+            .context(error::SortArraysSnafu)?;
+        let op_type = arrow::compute::take(&op_types.to_arrow_array(), &row_selection, None)
+            .context(error::SortArraysSnafu)?;
+
         let batch = self.adapter.batch_from_parts(
             key_columns,
             field_columns,
-            Arc::new(sequences),
-            Arc::new(op_types),
+            Helper::try_into_vector(sequences)
```
Comment (evenyag) [inline src/storage/src/memtable/btree.rs:225]: issue: If all rows are filtered then we might return an empty vector. I remember that batch_from_parts doesn't expect empty vectors.

File: src/storage/src/memtable/btree.rs [hunk 11]
```diff
@@ -196,25 +204,53 @@ impl BTreeIterator {
...
+                .context(error::ConvertChunkSnafu { name: "sequence" })?,
+            Helper::try_into_vector(op_type)
+                .context(error::ConvertChunkSnafu { name: "op_type" })?,
         )?;
 
         Ok(Some(batch))
     }
 }
 
+fn filter_keys(keys: &[&InnerKey], range: Option<&TimestampRange>) -> UInt32Array {
```
Comment (evenyag) [inline src/storage/src/memtable/btree.rs:235]: We have a `MapIterWrapper` to filter the entries by sequence. Maybe we can filter the entries by time range (if any) in it.

https://github.com/GreptimeTeam/greptimedb/blob/4ae0b5e1855ba37e0007e9f0c55fc5577b200c06/src/storage/src/memtable/btree.rs#L260-L268

**Pull request merged**